feat(SemVer): Add Semantic Versioning docs to repo and link it from i… (#103)
Shift3/standards-and-practices: git-and-github/code-versioning.md

# Git Tagging & Releases

## What is Git Tagging?

  * [Tagging](https://git-scm.com/book/en/v2/Git-Basics-Tagging) is a simple, easy way to keep track of your code at a certain point in time. Some instances where this becomes useful are:
    * Client-facing deliverables such as milestones, UAT, or production launch/release

>     * Internal-facing code at a point that you or your team wants to demarcate as important

Reviewer: > demarcate

You don't gotta get all fancy words on us lol

    * Versioning of your code (such as following SemVer or other release formats)

>   * Because we are not following a strict rule or methodology, each of our teams can decide for themselves which method of tag naming they want to follow. You could go strict and version your code based on milestones or releases, or you could use human-readable tag names. In the end, the power is _yours_!

Reviewer 2: I would challenge us to find and commit to one way to do this so we aren't reinventing the wheel here with each project.

Reviewer 2: That being said, as a whole I like this version, and think it's a good place to start.

Author: @ryekerjh `major.minor.patch`

# Semantic Versioning (SemVer)

## What is SemVer?

Ever used NPM before? How about Homebrew, or RubyGems? Each of those and most other package management systems need a way of keeping track of what version their public codebase is on, what versions they have had before, and what version they are heading toward.

The three package management systems I mentioned above all use a system of organization called [**Semantic Versioning**](https://semver.org/) to maintain uniform control of their code and how it is recieved by thier users. In short, versioning helps code maintainers and users of open source code identify several aspects of the code in question.

## SemVer Basics:
  * Always have an up-to-date documented API (in our case, README.md for our projects or Wiki)
  * Increment your versions based on a **Major.Minor.Patch** format where:

>     * **Major** changes are breaking, non-backwards compatable changes to your code base.

Reviewer: ```suggestion
    * **Major** changes are breaking, non-backwards compatible changes to your code base.
```

Author: Gooood catch!

>     * **Minor** changes are new features or functionality that ius still backwards-compatable

Reviewer: ```suggestion
    * **Minor** changes are new features or functionality that is still backwards-compatible
```

    * **Patch** changes are bug fixes or hotfixes that only address issues and don't introduce anything new.
  * Releases should be unalterable and locked -- forever.
  * Initial development (so most of our lives) should be treated as 0.x.x so as to indicate that there is no release as yet.
    * Generally, any software with a major version of 0 should be treated as unstable.

## Why should Shift3 care?

This should be self-evident: more maintainable code makes our jobs easier! Semantic Versioning does the following things  
  * Provides markers for points in our development cycle such as releases, patches, hotfixes, etc.
  * Keeps in-development projects organized by allowing for pre-releases (ie. 0.1.0, 0.2.0) that can be mapped to specific milestones
  * Allows for quicker set-up time for new developers being added to your project

>   * Help us avoid introducing major breaking changes unintentionally for enforcing the major/minor/patch methodology.  

Reviewer: ```suggestion
  * Help us avoid introducing major breaking changes unintentionally by enforcing the major/minor/patch methodology.  
```
